### ISSUE TITLE
Sorting is automatically reverted after any element in the listview is modified

### ISSUE THREAD

**Issue by darrengyl** [realm/realm-java#2604]
#### Goal

I am displaying a list of songs in the order of track number. Initially they are all displayed ASC. If I make a change to an individual song, all of them are still in the correct order. When I sort them by using the realm sort method, something weird happened. At first it was good that all songs are sorted in DSC order. But when I try to make a change to any song, the sorting order is reverted to ASC by track number.

I noticed that the listview was painted twice after the transaction is committed by checking getView method in adapter. It was first painted in DSC order, then something was called to set the order in ASC. I can't find what caused that. Please help me. The code itself is really simple though.

#### Expected Results

When I make change to the elements in the list view, my sorting won't be reset.

#### Actual Results

When I make change to the elements in the list view, my sorting is gone. However, no error is showing.

#### Steps & Code to Reproduce

 I have a list of songs.

I sort songs by id ASC by clicking on toolbar
I click on an item, then a int counter(not id) of this element will be incremented. I commit the transaction
Everything is fine.
I sort songs by id DSC  by clicking on toolbar again
I click on an item, then a int counter(not id) of this element will be incremented. I commit the transaction
The entire list is sorted again by id ASC, which is not expected. I debugged everything I wrote and nothing about sorting was executed.

#### Code Sample

MainActivity:

    public class MainActivity extends AppCompatActivity implements OnClickInterface {
        @Bind(R.id.tool_bar)
        Toolbar toolBar;
        @Bind(R.id.list)
        ListView list;
        private Realm realm;
        boolean isChecked = false;
        RealmListviewAdapter adapter;
        boolean sort = false;
    
        @Override
        protected void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            setContentView(R.layout.activity_main);
            ButterKnife.bind(this);
            setSupportActionBar(toolBar);
            realm = Realm.getDefaultInstance();
            adapter = new RealmListviewAdapter(this, null, true);
            adapter.setListener(this);
            list.setAdapter(adapter);
            if (isChecked) {
                updateAllSongs();
            } else {
                realm.where(Song.class).findFirstAsync().asObservable()
                        .filter(new Func1<RealmObject, Boolean>() {
                            @Override
                            public Boolean call(RealmObject realmObject) {
                                return realmObject.isLoaded();
                            }
                        })
                        .first()
                        .subscribeOn(AndroidSchedulers.mainThread())
                        .subscribe(new Action1<RealmObject>() {
                            @Override
                            public void call(RealmObject realmObject) {
                                if (!realmObject.isValid()) {
                                    addUsers();
                                } else {
                                    updateAllSongs();
                                }
                            }
                        });
                isChecked = true;
            }
        }
    
        @OnClick(R.id.tool_bar)
        public void onClickToolbar() {
            if (sort) {
                adapter.getRealmResults().sort("_id");
            } else {
                adapter.getRealmResults().sort("_id", Sort.DESCENDING);
            }
            sort = !sort;
            adapter.notifyDataSetChanged();
        }
    
        private void addUsers() {
            realm.executeTransactionAsync(new Realm.Transaction() {
                @Override
                public void execute(Realm realm) {
                    String[] titles = getResources().getStringArray(R.array.song_titles);
                    for (int i = 0; i < titles.length; i++) {
                        Song song = new Song();
                        song.set_id(i);
                        song.set_trackNumber(i + 1);
                        song.set_title(titles[i]);
                        song.set_favorite(false);
                        song.set_frequency(0);
                        song.set_lyrics("");
                        song.set_downloaded(false);
                        realm.copyToRealm(song);
                    }
                }
            }, new Realm.Transaction.OnSuccess() {
                @Override
                public void onSuccess() {
                    updateAllSongs();
                }
            });
        }
    
        private void updateAllSongs() {
            realm.where(Song.class).findAllSortedAsync("_id").asObservable()
                    .filter(new Func1<RealmResults<Song>, Boolean>() {
                        @Override
                        public Boolean call(RealmResults<Song> songs) {
                            return songs.isLoaded();
                        }
                    })
                    .first()
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(new Action1<RealmResults<Song>>() {
                        @Override
                        public void call(RealmResults<Song> songs) {
                            adapter.updateRealmResults(songs);
                        }
                    });
        }
    
        @Override
        public void onClickedItem(int position) {
            System.out.println("Pressed position " + position);
            Song song = adapter.getRealmResults().get(position);
            realm.beginTransaction();
            Log.d("frequency", String.valueOf(song.get_frequency()));
            song.set_frequency(song.get_frequency() + 1);
            realm.commitTransaction();
            Log.d("frequency", String.valueOf(song.get_frequency()));
            //Log.d("frequency", String.valueOf(_data.get(position).get_frequency()));
        }
    }

My adapter:

    public class RealmListviewAdapter extends RealmBaseAdapter<Song> implements ListAdapter{
        private OnClickInterface _listener;
    
        public RealmListviewAdapter(Context context, RealmResults<Song> realmResults, boolean automaticUpdate) {
            super(context, realmResults, automaticUpdate);
        }
    
        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            SongViewHolder viewHolder;
            if (convertView == null) {
                convertView = inflater.inflate(R.layout.eachsong, parent, false);
                viewHolder = new SongViewHolder(convertView);
                convertView.setTag(viewHolder);
            } else {
                viewHolder = (SongViewHolder) convertView.getTag();
            }
            Song song = realmResults.get(position);
            viewHolder.track.setText(String.valueOf(song.get_trackNumber()));
            viewHolder.title.setText(song.get_title());
            viewHolder.position = position;
            return convertView;
        }
    
        public void updateRealmResults(RealmResults<Song> queryResults) {
            super.updateRealmResults(queryResults);
            Log.d("Adapter", "UpdateRealmResults is called");
        }
    
        public RealmResults<Song> getRealmResults() {
            return realmResults;
        }
    
        public void setListener(OnClickInterface listener) {
            _listener = listener;
        }
    
        class SongViewHolder{
            @Bind(R.id.tracknumber)
            TextView track;
            @Bind(R.id.title)
            TextView title;
            @Bind(R.id.freq)
            TextView freq;
            @Bind(R.id.fav_star)
            ImageView fav;
            int position;
    
            public SongViewHolder(View itemView) {
                ButterKnife.bind(this, itemView);
            }
    
            @OnClick(R.id.item_row)
            public void onClickItem() {
                _listener.onClickedItem(position);
            }
        }
    }

If you are interested in it, the entire project can be found at https://github.com/darrengyl/realmSample

#### Version of Realm and tooling
Realm version(s): 0.88.2

Android Studio version: 2.1 Preview 5

Which Android version and device: 6.0.1 Nexus 5

**Comment by cmelchior** [comment]
Hi @darrengyl 
Thanks for reporting this. You are right that right now sorting isn't remembered correctly. We will look into it.

**Comment by kneth** [comment]
@darrengyl Sorry for the late reply.

As you can see in #2733, I have written an extra test. Recently we changed how `sort` was working. Instead of an in-place sorting method, `sort()` will now return a new `RealmResults`. The test in #2733 indicates that sorting order is preserved after a commit.

I have cloned your Github repository and updated to latest Realm (0.90.0). As we have moved the adapter out of Realm, of course I also had to add that. A few modifications of your app, and it seems to work as you except that is, preserving the sorting order. You can find my modifications at https://gist.github.com/kneth/d892973141e09fa12276ba7fc13c98f4

So long story short: we fixed a bug by changing how `sort()` is working, and you'll need to upgrade to made it work.
